docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+provides a check in `ValidatePodTemplateSpecForRC()`. Since *Running Debug* must
+be implemented server-side, a better long term strategy may be to migrate *Copy
+Debug* server-side as well to take advantage of easier validation against
+current cluster state.
+
+#### Stripping labels
+
+Only the spec will be copied from the running pod. Copying metadata such as
+labels might result in the pod copy receiving traffic or being killed
+immediately by a replication controller. It would be trivial to provide a
+`--copy-labels` option if this is desired behavior.
+
+### Running Debug Mode
+
+*Running Debug* mode is a `kubectl debug` invocation that results in a new
+container being introduced in the running pod. This *debug container* is not
+part of the pod spec, which remains immutable, and is not restarted according to
+the pod's `restartPolicy`.
+
+The status of a *debug container* is reported in a new `DebugContainerStatuses`
+field of `PodStatus`, which is a read-only list of `ContainerStatus` that
+contains an entry for every *debug container* that has ever run in this pod.
+Additionally, a new boolean `Tainted` in `PodCondition` changes from false to
+true upon execution of a *debug container*. `Tainted` could also be flipped by
+other operations which change a pod to a non-pristine state, such as `kubectl
+exec`. A transition of `Tainted` from true to false is not possible.
+
+Debug operations will generate an event so auditing utilities can reconstruct
+what commands are run, though `kubectl exec` does not do this. A *Debug
+Containers* section will display all debug containers in the output of `kubectl
+describe` similar to the *Init Containers* section, and the status of `Tainted`
+will be displayed in the *Conditions* section.
+
+The following command would attach to a newly created container in a pod:
+
+```
+kubectl debug -p target-pod -it -c debug-shell --image=debian -- bash
+```
+
+It would be reasonable to provide a default container name and image which,
+combined with a default entrypoint, makes the minimal debug command:
+
+```
+kubectl debug -p target-pod -it
+```
+
+If the specified container name already exists, `kubectl debug` will either
+attach to an already running container or restart and attach a container that
+has exited. Specifying arguments is not allowed in the first case but is allowed
+in the latter to support remote shell uses cases such as `kubectl debug -p
+target-pod -- netstat`.
+
+#### Additional kubelet complexity
+
+Implementing debug requires no changes to the Container Runtime Interface as
+it's the same operation as creating a regular container. The majority of the
+complexity for this feature falls to the kubelet, but it's simplified by the
+lack of configuration and restart policy. The kubelet's additional
+responsibilities include:
+
+1.  Start or restart a debug container in a pod when requested.
+1.  Don't kill the debug container while the pod is alive.
```

Comment:
@verb IIUC, since you treat the 'debug' session similar to the 'exec' session, the debug container will terminate whenever the command finishes, or the user disconnects? In other words, there should not be a running debug container with an active user streaming session. Is that correct?